### PR TITLE
HAL refactoring: HighPriorityTrigger and SysControl

### PR DESCRIPTION
**Pull request by martukas**
This refactors the HAL some more, in particular abstracting `HighPriorityTrigger`, `SysControl` and improving `Timers` and moving more stuff out of `hal_stm32.*`

Updates #1184

## Self-review checklist:

- [x] Self-review: looked through the `Files changed` tab, browsed repository in branch
- [x] Documentation updated - reflects changes in code, electrical or mechanical design
- [x] Documentation and graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] New content is linked, easily discoverable, does not require too many clicks
- [x] Follows other relevant parts of [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki)
- [x] PR has a descriptive name
- [x] Tagged relevant reviewers

### Software only:

- [x] Follows our [code style](https://github.com/RespiraWorks/Ventilator/wiki/Code-style)
- [x] Commented code, particularly in hard-to-understand areas
- [x] All tests pass - new and old, locally and on CI
- [x] No new warnings or static check failures
- [x] Tests that prove fix is effective or new feature works
- [x] Manual tests are explained, with instructions for reproducing them

### CODE REVIEW

File: software/controller/lib/core/system_constants.h [hunk 1]
```diff
@@ -23,10 +23,32 @@ limitations under the License.
...
+ * as PWM outputs:
+ *
+ * PA8  - Timer 1 Channel 1 - heater control
+ * PA11 - Timer 1 Channel 4 - psol control
+ * PB3  - Timer 2 Channel 2 - blower control
+ * PB4  - Timer 3 Channel 1 - buzzer control
+ *
+ * These timers are documented in [RM] chapters 26 and 27.
+ *****************************************************************/
+
```
Comment (asmodai27) [inline software/controller/lib/core/system_constants.h:51]: Mappings are redundant with the individual comments on each PwmChannel, and actually the documentation in [RM] is related to the timers, and belongs in the pwm implementation which sets the timer registers.

File: software/controller/lib/core/v03system.cpp [hunk 2]
```diff
@@ -133,6 +144,10 @@ void V03Actuators::execute(const ActuatorsState &desired_state) {
   }
 
   psol_->set(desired_state.fio2_valve);
+
+  desired_state.led_red ? led_red_->set() : led_red_->clear();
+  desired_state.led_yellow ? led_yellow_->set() : led_yellow_->clear();
+  desired_state.led_green ? led_green_->set() : led_green_->clear();
```
Comment (asmodai27) [inline software/controller/lib/core/v03system.cpp:150]: I'm not convinced that setting the LEDs belongs in `actuators.execute()`: we may want to do stuff with the LEDs that could be asynchronous from the main control loop, much like the buzzer.

File: software/controller/lib/hal/hal_stm32.cpp [hunk 3]
```diff
@@ -83,40 +57,27 @@ extern "C" void abort() {
...
 void HalApi::EarlyInit() {
-  // Enable the FPU.  This allows floating point to be used without
-  // generating a hard fault.
-  // The system control registers are documented in [PM] chapter 4.
-  // Details on enabling the FPU are in section 4.6.6.
-  SysControlReg *sys_ctl = SysControlBase;
-  sys_ctl->coproc_access_control = 0x00F00000;
-
-  /// \TODO initialized but never used. Do we still need it?
+  SysControl::enable_fpu();
```
Comment (asmodai27) [inline software/controller/lib/hal/hal_stm32.cpp:66]: Maybe keep the rationale (first two lines of that comment) as to why we enable the FPU here, even if the part about registers move to where they are abstracted away?

**Review by asmodai27 — APPROVED** [review]
Some minor remarks, but that looks like a nice improvement.

File: software/controller/lib/core/v03system.cpp (hunk 2 above)
Comment (martukas) [reply software/controller/lib/core/v03system.cpp:150]: Whenever we get around to using them, it should be easy enough to change this aspect of the interface. But as instances they do belong in this class. Also, I believe they are somehow not fully functional in v0.3. I recall Edwin saying some of them were places on the board upside down and are not visible. So I'm gonna leave this as is and if/when we need it, we can improve this.

File: software/controller/lib/core/system_constants.h (hunk 1 above)
Comment (martukas) [reply software/controller/lib/core/system_constants.h:51]: ok

File: software/controller/lib/hal/hal_stm32.cpp (hunk 3 above)
Comment (martukas) [reply software/controller/lib/hal/hal_stm32.cpp:66]: ok

**Pull request merged**